### ISSUE TITLE
Adding Kerberos support to Docker

### ISSUE THREAD

**Issue by dimastopel** [moby/moby#13697]
Hi Guys

We develop a container security suite oriented to the enterprises. As part of it we support Kerberos as an authentication mechanism between Docker CLI and the Docker daemon. We currently have it up and running but, honestly, think it should be part of the platform. Hence, we would like to contribute the code to Docker. 

Today we use Nalin's GSS Go wrapper (https://github.com/nalind/gss) over vanilla MIT Kerberos (https://github.com/krb5/krb5). 

I'd like to propose the following, simple, addition to the command line interface that enables generating Kerberos ticket (TGS) on the client (CLI), and require validating Kerberos ticket on the daemon.

CLI example:

```
docker -H docker-host:8080 --kerbverify --kerbspn=<service_name> ps
```

Docker daemon example:

```
docker -d --kerbverify --kerbspn=<service_name / *> --kerbkeytab=/tmp/docker_keytab
```

The SPN argument is mandatory. However, keytab argument is not. In case keytab is not specified, a default path is used. The default path is: `/etc/krb5.keytab` (as in [MIT krb5](http://web.mit.edu/KERBEROS/krb5-1.12/doc/mitK5defaults.html))

All variables can be configured using environmental variables:

```
DOCKER_KERB_VERIFY
DOCKER_KERB_SPN
DOCKER_KERB_KEYTAB
```

We will support MIT krb5 and Microsoft Active Directory as KDC.

I would be glad to hear you general thoughts and if make sense I'll send a more elaborated technical descriptions of what we'd like to do.

**Comment by duglin** [comment]
@dimastopel is the ticket sent to the daemon via an http header?

**Comment by dimastopel** [comment]
@duglin Yes, HTTP Authorization header

**Comment by duglin** [comment]
@dimastopel you may want to look at how the config file allows you to specify additional http headers for all CLI->Daemon flows:  https://github.com/docker/docker/blob/master/cliconfig/config.go#L43
This was specifically added for these types of authentication usecases.  For our use, we have a "login" type of command (separate exe) that will populate the config file with the headers we need, then the user can do normal docker commands w/o needing to do anything special, like add new flags to each command.

Ideally, we'd like to have a CLI plugin do that and have `docker login ...` be what the user sees.
See #13140

**Comment by NathanMcCauley** [comment]
@dimastopel this is a good direction to consider, I'd be interested in more elaboration on the technical design behind this. Especially interested in the daemon side handling of the ticket and access control that would be based on the ticket there.

**Comment by dimastopel** [comment]
@NathanMcCauley sure thing. We are working on this and I'll share a more elaborated story in the next couple of days. 

@duglin thanks. Will check this out.

**Comment by dimastopel** [comment]
@NathanMcCauley

Below is a more technical elaboration on how I see each component will operate. It doesn't contain the integration points yet. 

#####Client

In case `--kerbverify` is specified and SPN is provided, the following steps will be performed within the CLI client:

1. Obtain the current user credentials. This is basically the Ticket Granting Ticket (TGT) that already exists on the system.
2. Obtain the ticket to authenticate against the service. This is the Ticket Granting Service (TGS). Unless cached, obtaining TGS requires contacting the Key Distribution Center (KDC). The communication is transparent to the user. 
3. Encoding TGS using base64 as sending as a payload via HTTP Authorization header to the daemon as part of the request. 

In case of any error during the process, the client exits with the appropriate error code and message. 

Example request header with the token:

    Authorization: Negotiate YWRranNsY...==

#####Daemon

Upon receiving the request the server extracts the TGS from the HTTP Authorization header and validates it. No communication with KDC is required to validate the TGS. Once the TGS is validated, and the User Principal Name (UPN) is extracted the authentication process completes. 

In case the authentication process fails an HTTP 401 code is returned to the client. In case the SPN was incorrect HTTP 403 will be returned. 

Kerberos authentication requires one hop between the client and the daemon. 

#####Access Control

We use authentication both for auditing but also to provide the ability to create policies that control user access. For example, limit each developer to 5 containers max on the specific server. See below an example screen for configuring such policy. 

I propose a thin framework that will enable adding external modules via Docker HTTP API. The modules are external services that expose HTTP REST API. The framework will enable registering to system events, so that each module will be able to register for the relevant event groups. However the framework will also enable the modules to _control_ the flow of the system via return codes. This implies that the calls to the modules will be blocking unless (configurable) timeout occurs. In our case we will register for _user authenticated_ event and check against the policies we have whether this specific user is authorized to perform the specific task. If we see that the user is not allowed to do the task we need to be able to tell the daemon to stop processing the request and provide the appropriate reason. 

I didn't investigate yet to what extent the current events infra can be used to achieve this. I guess the current events infra is very relevant. Conceptually, this is also close to what @duglin suggests for CLI.

I think the line between the platform (Docker) and an external partner (e.g., Twistlock) should be this very framework. The framework is part of the platform but the modules are supplied by the external partner. As part of the partner's installation process the module adds itself via Docker's daemon (future) API.

Last note on this. In my opinion, we should differentiate between Kerberos authentication support and access control, and concentrate first on Kerberos authentication only. Once we have this we would be glad to go further with Access Control support. We would be glad to be the once to drive and implement both. 

![Docker Policy Example](https://www.dropbox.com/s/o19hb8r52z6ilru/policy_docker_1.png?dl=1)

Your comments on this are very welcome.

**Comment by thaJeztah** [comment]
I realise this is less important for the global concept of this proposal, but I would prefer to use a more generic flag to pass the options, similar to the way options for logging and graph-drivers are passed. That way other authentication mechanisms can be added in future without introducing new flags.

Something like;

    docker -d --auth-opt kerb-verify=... --auth-opt kerb-spn=...

(perhaps even `kerberos-verify=...`)

>  we should differentiate between Kerberos authentication support and access control, and concentrate first on Kerberos authentication only.

Fully agree on that; access control should not be tied to authentication here.

I :heart: the looks of that screenshot though!

**Comment by dimastopel** [comment]
@thaJeztah - fully agree with your proposal re flags. 

If there are no major objections to what I have proposed I’m going to draft a pull request with the initial implementation and continue the discussion from there. 

If there are objections please let me know ASAP.

**Comment by thaJeztah** [comment]
> If there are no major objections to what I have proposed I’m going to draft a pull request with the initial implementation and continue the discussion from there.

Purely based on the discussion above, I don't think there are objections (but please don't hold that against me :)). Creating a draft pull-request usually works well, as that gives the maintainers a bit more insight in the proposed changes.

**Comment by NathanMcCauley** [comment]
@dimastopel Have you considered what it would look like for this notion of pluggable authentication to be implemented as integration with PAM? Then, a bunch of the existing work on different PAM backends could be utilized by the Docker daemon to allow varied methods of authentication (LDAP, kerberos, OTP, etc).

**Comment by dimastopel** [comment]
@NathanMcCauley Let me look into it.

**Comment by nalind** [comment]
My guess is that, if we were to call into PAM at all, it would be for implementing the password checking part of the Basic auth scheme at the server.

**Comment by dimastopel** [comment]
Sorry for the late reply on this

@NathanMcCauley

After researching this for some time I agree that PAM looks as a great way to get "pluggable" behavior with a relatively low effort. In addition, numerous existing PAM modules can be used OOB. 

It also became clear to me that adding generic PAM support to Docker is a superior task than adding a vanilla Kerberos support. Specifically with Kerberos, PAM might not provide a good solution to the Single Sign On (SSO) functionality, but this issue can be addressed after PAM is in place. [This module](http://www.eyrie.org/~eagle/software/pam-krb5/), for example, claims it solves this issue OOB. I didn't verify it yet, though.

**Unless there are objections we will work on the general PAM support and then move to Kerberos.**

@nalind

As far as I understand we can get support for more than just basic AuthN in case PAM is in place. E.g., [OTP](http://www.cl.cam.ac.uk/~mgk25/otpw.html), [LDAP](http://www.padl.com/OSS/pam_ldap.html), [Challenge-Response with Yubico](https://developers.yubico.com/yubico-pam/Authentication_Using_Challenge-Response.html), etc. What do I miss?

**Comment by thaJeztah** [comment]
ping @icecrime to check if this lines up with the roadmap

Thanks in advance, @dimastopel !

**Comment by nalind** [comment]
@dimastopel PAM's generally only going to be of use when you want to ask for information that can be directly input by a user, such as a password which can be used for obtaining Kerberos credentials, and Basic auth more or less works for that case.  The PAM model allows modules to prompt and ask other questions, too, which is how it can support things like OTP.  When modules start asking for things other than passwords, or asking for more than one piece of information, the limitations of what you can communicate between the client and the server using the Basic scheme start creating problems -- a server implementation that just supplies the user's Basic password whenever modules ask for information during authentication is likely going to be providing the wrong answer to some of those modules.

Going by OpenSSH as an example, full PAM support, which avoids that sort of problem, is only available when using "keyboard-interactive" authentication, while "password" authentication's use of PAM is there as a best-effort for the sake of clients that don't implement "keyboard-interactive".  When authentication is being done using a binary wire protocol like the one used in GSSAPI and/or Negotiate, or using public keys, PAM isn't used for the authentication step, and the daemon only calls into PAM's account management function to do an authorization check.

In #13994, which is linked here because it overlaps somewhat, I've been focusing on making client- and server-side implementations of auth schemes pluggable, in a way that would allow the server's Basic implementation to be swapped out, or to have more than one available, calling into all of the built plugins which don't disable themselves until one of them succeeds.  On a second branch that builds on that, there's a Basic server implementation that uses libsasl2's sasl_checkpass() function.  The libsasl2 library can be configured to send the user names and passwords off to saslauthd, and saslauthd can be configured to use PAM to check them.  An implementation that calls directly into libpam could be slotted in next to it.

**Comment by icecrime** [comment]
Nothing conflicting with the roadmap from my point of view. I'll leave it to @NathanMcCauley and @diogomonica to see how they want to move forward with this.

**Comment by NathanMcCauley** [comment]
I support implementing an integration with PAM. Having PAM available for authn cases as well as available for authz for the key-based authentication schemes makes sense.  @nalind's #13994 looks like a step in the right direction for hooking up client<->daemon.

**Comment by dimastopel** [comment]
@nalind thanks for the reply!

Can you please elaborate on "_a server implementation that just supplies the user's Basic password whenever modules ask for information during authentication is likely going to be providing the wrong answer to some of those modules._"? Not sure I understand why server supplies authn information and why wrong answers will be provided. 

Regarding text based authentication, as far as I know, PAM supports binary messages and client side modules that enables quite rich customization of the authentication process. 

As I understand, your solution provides a plug-in based approach for the authentication of the Docker client against the Docker daemon. I think correct? PAM provides something that is quite similar. Do you see your solution as an alternative to PAM? If not, how do you see the two solutions working together?

**Comment by nalind** [comment]
@dimastopel The traditional strategy people have used when trying to support PAM in implementations of protocols like FTP is to call into PAM with a prompter callback that walks the list of passed-in prompts and hands back the user name when the msg_style is PAM_PROMPT_ECHO_ON, and to hands back the password when the msg_style is PAM_PROMPT_ECHO_OFF.  You run into weird behavior and bug reports if the dialog needs to be more complicated than that.  If the prompt text is required in order for the user to supply the right information (I'm thinking of challenge-response OTP schemes here), and it isn't displayed, or if the user is asked for information before the PAM conversation even starts, it can't work.  That's a risk when trying to fit something like PAM into a wire protocol that can't be as flexible as PAM expects it to be.  OpenSSH ended up needing to implement keyboard-interactive authentication to be able to properly handle the cases where its password authentication scheme wasn't flexible enough.

The Linux-PAM binary prompt interface isn't widely adopted.  Outside of the Linux-PAM tree itself, I see one module (the Openwall pam_userpass module) making use of the constants defined as part of the interface, and no references to it in the source trees of the packages in Fedora that link to libpam.  Granted, that's not exhaustive.

The patch I'm proposing makes things pluggable at the level of client and server implementations of authentication schemes - client and server implementations for Basic and Negotiate (or others, though I haven't tried to implement others) can be enabled (or not) at build-time, and the internal interface allows them to disable themselves at startup (for example, if the Negotate server discovers that it has no creds, so that client's won't be told that they can try Negotiate when the server is unable to accept it).  The server portion of Basic would be where I'd expect to be calling into libpam.  Negotiate, for example, would call into a GSSAPI library.

My personal leaning toward using libsasl2 for the server portion of Basic is based partly on wanting to avoid having to implement that sort of logic in a conversation callback, partly on liking that the process separation we get by having libsasl2 call out to the external saslauthd process reduces possible side-effects, and partly on saslauthd offering a caching option.

**Comment by dimastopel** [comment]
@nalind Makes sense. 

Several questions:

1. libsasl2 [seems to support](https://cyrusimap.org/docs/cyrus-sasl/2.1.25/readme.php) GSSAPI and plenty of other authN mechanisms. Why not to use it for everything, including Negotiate, and not just the Basic part? 
2. In case libsasl2 is used for the Basic part, it can be configured to use PAM (for those who are interested) *without modifying Docker at all*. [This](http://archives.neohapsis.com/archives/postfix/2007-09/att-0505/README-Postfix-SASL-RedHat.txt) is one example of such setup. Do I understand this correctly?
3. You mention that you plan to enable / disable authN methods at build time. Why not in run time via the configuration?

**Comment by nalind** [comment]
@dimastopel Thanks for suffering through my wall of text!

Using libsasl2's GSSAPI support would require some additional pipe-fitting -- to start, we'd have to map from HTTP authentication schemes to SASL mechanisms.  While that part looks pretty straightforward for Basic and Negotiate, I don't have a good plan for what we'd do for handling Bearer or OAuth, as there doesn't appear to be a simple mapping between the two sets (going off of http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml and https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xml).  The SASL GSSAPI mechanism specifies Kerberos as the only supported GSSAPI mechanism, while Negotiate specifies SPNEGO, and that could create difficulties if the server side is, or ever becomes, clever enough about validating incoming data.

You read correctly, saslauthd can call into PAM to check the password, though you have the same limitations there that you'd have when calling into PAM directly - the protocol between libsasl2 and saslauthd is a single round trip (the library sends the user, password, SASL realm, and service name, and receives either an "OK" response or an error message), but I hope it's clearer to .

If/when we disable mechanisms at build-time is more of a question of how we're handling library dependencies - if we're trying to gracefully handle the absence of some libraries at build-time, then we can have the build scripts check for the presence of libraries using pkg-config, and set build tags to build things conditionally.  If we're not, and we can straight-up require libraries to be present to build, we can build everything in.  Either way, the various schemes can check the run-time configuration and either not register themselves at startup, or not participate on a per-request basis.  We don't know whether or not libsasl2 and saslauthd are even set up at run-time, and offering a scheme that can't possibly succeed is going to annoy clients (I'm thinking particularly about prompting for passwords, only to have the password check fail every time), so we'd a runtime option regardless.

**Comment by dimastopel** [comment]
@nalind Thanks

PAM limitations are clear. What was important for me to verify is that customer with an existing PAM infra (or who just prefers using PAM) will be able to use it with Docker+libsasl2. And as far as I understand the answer is yes. 

I think that we can figure out along the way how to deal with issues like SASL + Bearer, OAuth.

Trying to take a bit more pragmatic approach here: @lioryanko from our team tried to build nalind/docker/daemonauth so we could play with it, identify the gaps, and try to build a mutual plan on closing those gaps. What we see is that some of the ```+build``` constraints of authenticators aren't met and they aren't compiled. What is the right way to compile the code?

Basically I think that we need to start with two chunks: 1) add pluggable authN framework with (indirect) PAM support. 2) Add support for Negotiate. Then we can continue with OAuth etc. 

WDYT?

**Comment by nalind** [comment]
Right, the `authframework` branch is just the framework patch, while `daemonauth` has that and some things that build on top of it.  In the `daemonauth` branch, the Basic and Negotiate bits are currently conditionalized on whether or not the build scripts detect libsasl2 and the GSSAPI libraries by running pkg-config at build-time, and that we're not static linking, since we don't build Kerberos or Cyrus SASL as static libraries in Fedora.  If we have to work around that last part to make things available in static builds, we can deal with that.  FWIW, for quick checks, I tend to run 'go build -tags "daemon gssapi libsasl2"' from the top-level 'docker' subdirectory.

**Comment by lioryanko** [comment]
@nalind Your code looks good. I'm trying to build it now.

I'm having a bit of trouble building your daemonauth branch (without excluding the server's authenticator modules). When trying to build the way you said (from inside docker's build environment container, built using the Dockerfile which includes your modifications) I got the following errors:
```
# github.com/docker/docker/daemon/graphdriver/btrfs
../daemon/graphdriver/btrfs/version.go:6:27: fatal error: btrfs/version.h: No such file or directory
 #include <btrfs/version.h>
                           ^
compilation terminated.
# github.com/docker/docker/pkg/devicemapper
could not determine kind of name for C.dm_task_deferred_remove
```

I tried using 'go build -tags "daemon libsasl2 gssapi btrfs_noversion libdm_no_deferred_remove"' (hope that makes sense). The error I'm having now is:
```
# pkg-config --cflags libsasl2 krb5-gssapi
Package libsasl2 was not found in the pkg-config search path.
Perhaps you should add the directory containing `libsasl2.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libsasl2' found
pkg-config: exit status 1
```
From what I can tell it looks like libsasl2-dev package supplied by ubuntu is outdated and doesn't include a package configuration file. 

Any thoughts?

**Comment by nalind** [comment]
@lioryanko It looks like you need a newer version of the `btrfs-tools` package to get `/usr/include/btrfs/version.h`, too.

As for libsasl2, replacing `#cgo pkg-config: libsasl2` with `#cgo LDFLAGS: -lsasl2` should do the trick, since the package places the headers in the default search path.  It's less future-proof, I guess, but it appears it's going to necessary for regular builds until we move to using 15.04 or something later as the default build environment.  I'll make that change shortly.

**Comment by lioryanko** [comment]
@nalind Well, the basic authentication looks good:

```
root@9253f757ae8a:/go/src/github.com/docker/docker# ./docker/docker ps
Username: fake       
Password: 
Error response from daemon: wrong login/password
root@9253f757ae8a:/go/src/github.com/docker/docker# ./docker/docker ps                        
Username: lior
Password: 
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

There are bugs though which I haven't really inspected thoroughly, here's an example:
```
root@9253f757ae8a:/go/src/github.com/docker/docker/docker# ./docker run -ti busybox /bin/sh
Username: lior
Password: 
Post http:///var/run/docker.sock/v1.20/containers/create: http: ContentLength=968 with Body length 0.
* Are you trying to connect to a TLS-enabled daemon without TLS?
* Is your docker daemon up and running?
```

Also, when trying to build an image (note that there's no authentication prompt):
```
root@9253f757ae8a:/go/src/github.com/docker/docker# ./docker/docker build -t inception_image .
Sending build context to Docker daemon 241.4 MBn 557.1 kB
```
All in all it looks pretty sweet :)

**Comment by nalind** [comment]
@lioryanko It looks like I missed the part about retransmitting the request body when we retry a request with authentication, which is complicated by the body coming from stdin some of the time.  Going to have to work on that.

**Comment by nalind** [comment]
Okay, that should work better.  The Basic client now assumes that passwords can be reused, so that operations like "run --rm" don't end up prompting for credentials a half dozen times.  Using a TeeReader and MultiReader to try to cache only part of the request body didn't work so well, so for now we read and hang on to the whole thing.  At least if we end up using too much memory, it's the client that's doing it, so the daemon doesn't suffer.

**Comment by aleemb** [comment]
Does this feature prevent `docker attach` and `docker exec` without authentication?

Reference: #14473.

**Comment by dimastopel** [comment]
Yes, definitely.

**Comment by lioryanko** [comment]
@ nalind Basic authentication works great as far as I can tell.

I also managed to authenticate using GSS-API (Kerberos). There are two things that I don't really get, though. First, why do you try to convert client principal names to local names (by using gss_localname)?
Secondly, it looks like you don't support GSS continuation (on purpose).

**Comment by nalind** [comment]
@lioryanko Using gss_localname() should give us a chance to allow a single user to authenticate using either Basic or Negotiate at different times, while being able to look like a single user.  I'm not wedded to it, though -- earlier revisions used the results from gss_display_name().

Supporting continuation might require maintaining more state on each end.  I'm not sure we end up talking to anything that could/would use it.  Are you hitting a case where it's broken without it, or is it more of a would-be-more-correct thing?

**Comment by lioryanko** [comment]
@nalind I can see the appeal of what you're saying regarding the single user for multiple schemes. The problem with that approach though is that it assumes that for each authenticated user there's a corresponding local user. I have an existing setup where this approach fails with kerberos authentication because KDC principals don't have matching local users per each host.

Well continuation kind of fails. With my setup I'm getting a warning for each authentication, saying continuation is needed. I also get a token that needs to be returned and processed by the client. In addition, it seems more future-proof.

If that's OK with you I'd be happy to make those changes myself.

**Comment by jerbia** [comment]
@lioryanko @nalind The need to map between KDC principals and local users raises a question regarding user authorization.
How can we prevent non-admin users, with valid kerberos ticket, from managing containers?
Up until now the approach for authorization was one of the following:
1. Users need to be member of local 'docker' group
2. Users need to have a valid certificate. 

These methods are somehow "controlled". 
Mapping KDC to local user will probably mean that local user need to be member of 'docker' group.
If we don't require mapping then we need some other access control mechanism to prevent users with kerberos ticket (e.g. every Active Directory user) from issuing Docker commands.
This means that we will need to have some way for specifying which users are allowed to manage Docker.

Any thoughts on that?

**Comment by dimastopel** [comment]
@jerbia

Currently, there is no mechanism to _authorize_ the operations of authenticated, _remote_, users against Docker daemon. There is a client certificate authentication that can be used to differentiate between the uses who can access Docker with full privileges and those that can't access at all. The authentication framework we are trying to push (including Kerberos) should work the same in that sense: either you can get the TGS for the Docker service or you don't. It will support a bunch of additional authentication schemes but nothing authorization related. 

As discussed with @NathanMcCauley and @thaJeztah above we will tackle authorization in a separate effort. I've opened a separate issue this. Please see #14674.

**Comment by nalind** [comment]
@lioryanko If you mean using the display name instead of mapping to a local name (which, to be clear, produces a string, which isn't being validated any further than that), sure.  The thing about having the client process the server's token, which is probably the second leg of a mutual authentication, is that it accompanies the result of the client's request, so you have to have a plan for what to do if it isn't accepted.  Anyway, what were your thoughts on keeping the client state around?

**Comment by lioryanko** [comment]
@nalind Yeah I meant display name. The mapping did fail though when I tried to use a complex principal name (such as "lior/admin@docker" instead of "lior@docker"). 
The ret flags returned by gss_init_sec_context are GSS_C_CONF_FLAG|GSS_C_INTEG_FLAG|GSS_C_TRANS_FLAG, no mutuality according to that.
If continuation is needed simply for mutual authentication then it's pointless right now, as the client doesn't even require the server to authenticate, it's the server which is enforcing authentication. If the client would have enforced authentication then mutual authentication obviously would have made much more sense.
As for keeping the client state, I think that using http://www.gorillatoolkit.org/pkg/sessions would make it pretty straight forward...
What do you say?

**Comment by nalind** [comment]
@lioryanko The default mapping rule in krb5.conf only handles principal names with one component in the default realm, so yeah, that'd break without custom configuration or logic.

I was mistaken in thinking that a reply would only be needed for mutual auth - it's part of the SPNEGO flow, too.  Otherwise, since we're not requesting GSS_C_MUTUAL_FLAG, we could be getting back GSS_S_COMPLETE rather than GSS_S_CONTINUE_NEEDED when we first call gss_init_sec_context.

gorilla/sessions would probably be a reasonable way to keep track of in-progress context establishment operations in the server, so long as we don't end up unintentionally trying to carry those context IDs across restarts, and it would certainly make things faster for clients that make multiple connections to the daemon in quick succession.

**Comment by elric1** [comment]
Couple of quick comments about GSSAPI support:

``The SPN argument is mandatory. However, keytab argument is not. In case keytab is not specified, a default path is used. The default path is: /etc/krb5.keytab (as in MIT krb5).''

Maybe it would be better to make the service principal's name optional rather than mandatory.  Both MIT krb5 and Heimdal will accept any principal defined in their configured keytab if a name is not specified and so it should not be necessary.  In fact, it merely provides information that can already be determined from the keytab and in so doing you obtain another opportunity to misconfigure things for no benefit.  Hard-wiring to a single service principal name also effectively disables the ability to have one listener receive connexions to multiple addresses as the client will mostly choose the name based on the address used to connect to the service.

I think that it would be more convenient if the client did not require a flag to do GSSAPI but rather just tried to do it if the server asks for it in its 401 response.  If you do not have credentials it will fail instantly and without fuss.  If you do have Kerberos creds, well, you are probably a Kerberos shop---and the server is configured to ask for them...  why not just give it a try.

Same for the server.  Rather than requiring a flag to enable Kerberos, why not just do a gss_acquire_cred() and turn it on if there are credentials available.

DOCKER_KERB_KEYTAB is unnecessary.  Both MIT krb5 and Heimdal already use an environment variable KRB5_KTNAME.  No need to create another one.  In fact, you might be able to elide the command line argument entirely under the assumption that the user can set KRB5_KTNAME or modify /etc/krb5.conf: [libdefaults] default_keytab_name.

**Comment by nalind** [comment]
@elric1 I agree with your comments.  The patch I have should differ from what you're describing only in that it offers a `--keytab` option whose value we pass to `gsskrb5_register_acceptor_identity` if one is specified.  You're right, though, that init scripts tend to offer a way to set `KRB5_KTNAME` in the environment, so that logic isn't strictly necessary, and it could be dropped out.  I don't have a strong preference there.

**Comment by lioryanko** [comment]
@nalind Well, it looks good to me. 
Have you written any tests yet? How about splitting up the work? I can write integration tests and leave unit tests to you or vice versa.

**Comment by nalind** [comment]
@lioryanko Getting authentication over multiple round trips working took a little longer than expected, but the Negotiate client should now be properly checking Negotiate data that it gets back from the server.  The server should be setting cookies after successful authentication, and the client should be using them.  I managed to get the client loop factored in a way that it only had to be written once, though at the cost of duplicating cookie data when we send it back to the server for non-hijacked connections.  I'm new to writing tests for go, so could probably use some advice there, particularly around mocking up some of the C APIs that we call into.

**Comment by lioryanko** [comment]
@nalind I don't know of any existing infrastructure to mock C functions from inside go. What makes most sense to be honest is to wrap GSS API and pass an interface which you could then mock manually/using gomock. 

I take it that I'll be writing the integration tests then?

**Comment by nalind** [comment]
@lioryanko If you're volunteering, sure! :smile: I have to spend the next little while working on other things, but I expect to be getting acquainted with writing tests next.

**Comment by thaJeztah** [comment]
Been following this discussion with interest. Compliments for the excellent work you're doing here. :heart: the way this is progressing :+1:

**Comment by rhatdan** [comment]
Yes this is great and very much needed...

**Comment by lioryanko** [comment]
Kerberos authentication integration tests PR: https://github.com/nalind/docker/pull/2

**Comment by dimastopel** [comment]
@NathanMcCauley @diogomonica  @thaJeztah  I think it is a good time now to start collecting your general feedback on the code in Nalin's branch. The work that takes place now is mostly around tests and it would be great to have an agreement, at least a general one, on the code before tests are finalized and proper PR is sent. What do you think?

**Comment by lioryanko** [comment]
Hey guys, there's a [google doc](https://docs.google.com/document/d/1lQr-9wnhd7yI7t8rF7cnB6hGGUZoYk6wvMO2ePLcx68/edit#) describing the design and UX of our work so far. Feel free to comment.

@nalind thanks for your help with this document :)

**Comment by ppadmavilasom** [comment]
Is there an auth plugin effort planned for docker in general? Trying to figure out if this is the right place to start.

**Comment by ppadmavilasom** [comment]
@lioryanko it seems you had success with basic auth. i can get htpasswd based auth to work. But so far no success on gssapi or libsasl. I ran the AuthSuite tests and they seem to pass. When i start with a --realm and connect over tcp, i get "An error occurred trying to connect: Failed to authenticate to docker daemon; server offered no authentication methods"
Any pointers?

**Comment by nalind** [comment]
@ppadmavilasom Try starting the server with debugging enabled.  The server doesn't offer Negotiate unless it can acquire acceptor creds at startup (i.e., unless it can read the keytab and there are keys in it), but it should log a debug message if that's the case.  If you're using mandatory access control, permissions may be a factor.

If the server is being started with --libsasl2 and --realm but isn't offering Basic, either, then there must have been a problem initializing the SASL library.  My best guess would be to check the /etc/sasl2/docker.conf or /usr/lib/sasl2/docker.conf files, both contents and permissions.

**Comment by cyphar** [comment]
I'm wondering whether this can just be implemented in the AuthZ plugin (#15365) and the lessons learnt making this could be channeled to improve the current AuthZ proposal.

**Comment by LK4D4** [comment]
@nalind @NathanMcCauley do you know what's status here? I see some references from another repos, but not sure how to interpret them.

**Comment by thaJeztah** [comment]
Closing this one because it went stale, and probably can be implemented using the AuthZ plugins
